pyoomph/__main__.py

```diff
@@ -101,97 +99,132 @@ def get_compilation_dir(parent_dir: pathlib.Path, name_hint: str) -> pathlib.Pat
-        all_compilation_units.append(unit)
-        unit.create_untyped_ast()
+        # Create a compilation unit out of it and parse it into an untyped ast
+        candidate_unit = CompilationUnit(source_path, session)
+        compilation_units.append(candidate_unit)
+        candidate_unit.create_untyped_ast()
 
-        deps = [
+        # Calculate its dependencies and add them to the dependencies dictionary,
+        # including builtins if necessary, and add those dependencies to the queue
+        current_dependencies = [
             top_declaration.path
-            for top_declaration in unit.ast
+            for top_declaration in candidate_unit.ast
             if isinstance(top_declaration, ast.Import)
         ]
         if source_path != project_root / "builtins.oomph":
-            deps.append(project_root / "builtins.oomph")
-        dependencies[unit.source_path] = deps
-        todo_list.extend(deps)
+            current_dependencies.append(project_root / "builtins.oomph")
+        dependency_graph[candidate_unit.source_path] = current_dependencies
+        queue.extend(current_dependencies)
+    return compilation_units, dependency_graph
+
 
-    # Dumbest toposort you have ever seen
+def compute_compilation_order(
+    verbose: bool,
+    compilation_units: List[CompilationUnit],
+    dependency_graph: Dict[Path, List[Path]],
+) -> List[CompilationUnit]:
     compilation_order: List[CompilationUnit] = []
-    while len(compilation_order) < len(dependencies):
-        unit = [u for u in all_compilation_units if u not in compilation_order][0]
-        decisions = [unit]
+    while len(compilation_order) < len(dependency_graph):
+        candidate_unit = next(
+            u for u in compilation_units if u not in compilation_order
+        )
+        breadcrumbs = [candidate_unit]
         while True:
-            need_first = [
+            uncompiled_dependencies = [
                 u
-                for u in all_compilation_units
+                for u in compilation_units
                 if u not in compilation_order
-                and u.source_path in dependencies[unit.source_path]
+                and u.source_path in dependency_graph[candidate_unit.source_path]
             ]
-            if not need_first:
+            if not uncompiled_dependencies:
                 break
-            if compiler_args.verbose:
+            if verbose:
                 print(
-                    f"Noticed dependency: {unit.source_path} --> {need_first[0].source_path}"
+                    f"Noticed dependency: {candidate_unit.source_path} --> {uncompiled_dependencies[0].source_path}"
+                )
+            candidate_unit = uncompiled_dependencies[0]
+            if candidate_unit in breadcrumbs:
+                message = (
+                    " --> ".join(d.source_path.name for d in breadcrumbs)
+                    + " --> "
+                    + candidate_unit.source_path.name
```

Comment:
I'm not sure if this produces the expected message when a file imports itself. I added a test to `main` branch in 50301a6. Can you check whether the test still passes?

Reply:
It works